### ISSUE TITLE
Add org.jooq.Generated annotation with RUNTIME retention for generated code

### ISSUE THREAD

**Issue by tazle** [jOOQ/jOOQ#8617]
### Use case:

We'd like to ignore generated code in our coverage measurements, but for jooq-generated code and jacoco that is currently not very easy.

Jacoco has recently gained support for ignoring generated code in coverage measurements, but only if a @Generated annotation with CLASS or RUNTIME retention is used. Currently jooq-generated code is annotated with javax.annotation.Generated, which only has SOURCE retention. Since Jacoco doesn't see the source files, it cannot use javax.annotation.Generated to ignore jooq-generated code.

### Possible solution you'd like to see:

Please use a custom @Generated annotation with CLASS or RUNTIME retention instead of / in addition to javax.annotation.Generated.

See https://github.com/jacoco/jacoco/pull/731 for Jacoco-side details.

### Possible workarounds:

It might be possible to use file name based exclusions for us.

### Versions:

- jOOQ: 3.11
- Java: 1.8
- Jacoco: 0.8.2 and newer

**Comment by lukaseder** [comment]
Thank you very much for your request. You can easily implement this feature by extending the `JavaGenerator.printClassAnnotations()` method.

I'll keep this issue open to see if it gets enough traction.

**Comment by knutwannheden** [comment]
Also, isn't JaCoCo able to exclude classes given regex patterns which are matched against fully qualified class names? Assuming that the jOOQ generated classes reside in Java packages where there is no other code, this might be an easier and arguably better solution.

**Comment by davsclaus** [comment]
Hitting this with Apache Camel now. That we do not have `javax.annotation.Generated` on the classpath out of the box in Camel 3.1 onwards. 

For Java 9 onwards there is a annotation out of the box `javax.annotation.processing.Generated` in the JDK itself. But for Java 8 users then yeah there is none.

**Comment by davsclaus** [comment]
This annotation are not in a good JAR from maven central. You can find it in some old glassfish JARs from 2009 :(

https://search.maven.org/search?q=fc:javax.annotation.Generated

**Comment by davsclaus** [comment]
Ah found it here

        <dependency>
            <groupId>javax.annotation</groupId>
            <artifactId>javax.annotation-api</artifactId>
            <version>1.3.2</version>
        </dependency>

**Comment by lukaseder** [comment]
@davsclaus The annotation should be picked automatically based on which JDK you're using to generate code. Otherwise, you can influence it in configuration as documented here:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-generate/codegen-generate-annotations/ (`generatedAnnotation` and `generatedAnnotationType`)

**Comment by lukaseder** [comment]
Notice, this issue is not about the package of the annotation, but about the retention...

**Comment by ericdriggs** [comment]
@lukaseder 
I was able to extend the JavaGenerator to use lombok.Generated, but it's not pretty -- in order to use it in my build.gradle, I needed to build it into a local jar to avoid the chicken and egg problem where all gradle Classloaders are resolved before any classes are compiled.

Would be helpful if you were to add a lombok generation option to jooq codegen similar to the example in the link below. I've verified that it generates the annotation and that jacoco ignores those classes (excepting static inner classes which currently don't get annotations).

https://github.com/ericdriggs/reportcard/blob/master/reportcard-jooq-generator/src/main/java/org/jooq/codegen/JavaLombokGenerator.java

**Comment by lukaseder** [comment]
Thanks for your suggestion, @ericdriggs.

**Comment by lukaseder** [comment]
This issue has a surprising amount of upvotes. I guess that means we'll add a `org.jooq.meta.jaxb.GeneratedAnnotationType.ORG_JOOQ_GENERATED` configuration for the `<generatedAnnotationType/>` configuration:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-generate/codegen-generate-annotations/

The annotation just mimicks the one at `javax.annotation.processing.Generated`, but with `RUNTIME` retention.

```java
package org.jooq;

import static java.lang.annotation.ElementType.CONSTRUCTOR;
import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.ElementType.LOCAL_VARIABLE;
import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.ElementType.PACKAGE;
import static java.lang.annotation.ElementType.PARAMETER;
import static java.lang.annotation.ElementType.TYPE;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * An alternative annotation to replace
 * {@link javax.annotation.processing.Generated} from the JDK, in order to
 * elevate scope to {@link RetentionPolicy#RUNTIME}, keeping an identical
 * structure.
 */
@Documented
@Retention(RUNTIME)
@Target({ PACKAGE, TYPE, METHOD, CONSTRUCTOR, FIELD, LOCAL_VARIABLE, PARAMETER })
public @interface Generated {

    /**
     * The name of the code generator.
     *
     * @see javax.annotation.processing.Generated#value()
     */
    String[] value();

    /**
     * The date the source was generated.
     *
     * @see javax.annotation.processing.Generated#date()
     */
    String date() default "";

    /**
     * Comments that the code generated included.
     *
     * @see javax.annotation.processing.Generated#comments()
     */
    String comments() default "";
}
```

**Comment by lukaseder** [comment]
Of course, you can always just search-replace generation output: `s/import javax.annotation.processing.Generated/import com.example.Generated/` as a workaround.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.19.0
